### PR TITLE
V8 WIP: Migrate DI abstraction to MS.DI

### PR DESCRIPTION
**Pull request by lars-erik**
### Description

Continuing #2690 to 
1. be able to run with MS.DI behind IRegister and IFactory
2. rename IRegister to IServiceCollection and IFactory to IServiceProvider
3. replace above interfaces with actual MS.DI abstractions
4. bundle the MS.DI Lightinject adapter

### Expected result

ASP.NET Core DI compatibility and freedom to use any container with Umbraco.

### State of PR

Everything is ready to start migrating. POC working with adapter from  
https://github.com/lars-erik/our.umbraco.containers

Discussion is going on in #v8-di on community slack.

### CODE REVIEW

**Comment by lars-erik** [comment]
Seems like controllers doesn't get properly auto discovered via adapter. I think it might be because I ignored auto-discovery for now.
Adapter is at: https://github.com/lars-erik/our.umbraco.containers

**Comment by nul800sebastiaan** [comment]
@lars-erik is this WIP or done? Please be aware that if you're doing more work and it requires breaking changes then that probably can't be done until v9 since we're trying to wrap up v8 asap.

**Comment by lars-erik** [comment]
@nul800sebastiaan WIP. I'll edit the header.
Wrap up V8 asap is good(ish), but don't let politics launch a half-baked solution.

**Comment by nul800sebastiaan** [comment]
Thanks! 

> don't let politics launch a half-baked solution. 
Care to elaborate?

**Comment by lars-erik** [comment]
Last suggestion to @zpqrtbnk was to implement IServiceCollection and IServiceProvider on IRegister and IFactory. If that interface is nailed, then we wouldn't need to do more breaking. (I think)
But it still doesn't work with MS.DI and if we _do_ want that, there just might be a few more changes needed before the interface can be considered "safe".
So don't let the suits launch V8 before it's technically ready, just because it looks good on the surface and peeps are nagging to get it early.

**Comment by lars-erik** [comment]
Sure, I know you have different forces working internally. As it should be. ;)  
Just hope I'm respected enough to be heard when I say: "It's not bloody ready!".
Ready for CTP / beta tho.

**Comment by JimBobSquarePants** [comment]
This is a priority IMO. Massive future proofing protection for V9 and beyond.

**Comment by lars-erik** [comment]
image

It now works with the current abstraction and the adapter I linked earlier.
I barely tested in the backoffice, so probably have a wrinkle or two, but POC is done!

We can start moving the MS.DI abstractions into Umbraco and push Lightinject down one notch in the dependency graph.

**Comment by lars-erik** [comment]
My current plan is do take it through these steps:

1. This is how it works today:
image

2. This is how it works with my adapter today:
image

3. This is an implementation that will mostly add and not remove anything from the current abstractions:
image

4. This is the optimal implementation. It still doesn't break anything implementers / package devs are using, since they should go through CompositionExtensions and Component.
image

Discussion is continuing in #V8-DI channel on community slack.

File: src/Umbraco.Core/Composing/LightInject/LightInjectContainer.cs [hunk 1]
```diff
@@ -84,41 +95,52 @@ protected static ServiceContainer CreateServiceContainer()
...
 
         /// <inheritdoc />
         public object GetInstance(Type type)
-            => Container.GetInstance(type);
+            => this.GetService(type);
 
         /// <inheritdoc />
         public object TryGetInstance(Type type)
-            => Container.TryGetInstance(type);
+            => this.GetRequiredService(type);
```
Comment (zpqrtbnk) [inline src/Umbraco.Core/Composing/LightInject/LightInjectContainer.cs]: Get and TryGet vs Get and GetRequired = should be the other way round

Comment (lars-erik) [reply src/Umbraco.Core/Composing/LightInject/LightInjectContainer.cs]: I suspected so. :) Thanks.

**Comment by JimBobSquarePants** [comment]
@lars-erik Hope you're feeling much better!

**Comment by lars-erik** [comment]
Thanks, @JimBobSquarePants :)
Been busy otherwise as well ofc.

I fixed the misunderstanding @zpqrtbnk pointed out above. Now down to 4 failing tests in that fixture.
Those fail because it actually "doesn't work that way" through LI's own adapter. I'm unsure whether it's due to config or conformance.

Speaking of conformance. The latest "targeted service" implementation seems a bit complex in order to hide the fact that we probably just require named dependencies.  
I looked at James' code and recognize it needs some polish to be used (OT), but there are several good ideas there.  

The question is: should we just require that MS.DI is used with an adapter that supports named services?

Listening to peeps on twitter, I'm not sure which end to start at. The work on abstracting LI has not really changed anything, except for a clearer interface and phases. It can still be even clearer, and filesystems, subsequently named services seems to be one of the remaining obstacles.

**Comment by nul800sebastiaan** [comment]
Hey everyone, as you know by now v8 has been released and additional changes to DI would be very breaking, which we're trying to avoid.

We are loving the passion and the efforts related to this subject and look forward to continue to iterate on this subject in v9 when we start working on that. Moving forward, let's pick this up again when we start a v9 branch, we'll let you know when that happens.

I also want to say that while I personally am not all that familiar with the underlying architecture, it's been a joy trying to use DI to the point that we have it now in v8 and it's also been a joy to see other people successfully using it. So a heartfelt thanks for all the efforts so far and there's more great updates to look forward to in v9. #h5yr

**Comment by JimBobSquarePants** [comment]
Super disappointed in this. Incredibly naive decision shipping before it was implemented

**Comment by lars-erik** [comment]
Thanks for the good feedback, @nul800sebastiaan!

I'm obviously a bit disappointed as well, but I respect the decision. I have also not been able to slot any more time for this since I've been prepping lectures all Q1.  
Have a slight hope that we can still squeeze this into a minor, but if I'll be the one to do it it will probably not happen until fall.  
In any case, the best thing we can do to leave the door open is manage to educate peeps on best practices, and by god, prevent them from adding the statement `using Lightinject` anywhere in packages or larger solutions.

**Pull request closed without merging**